zulip/zulip/__init__.py

```diff
@@ -652,10 +655,14 @@ def end_error_retry(succeeded: bool) -> None:
                     continue
                 else:
                     end_error_retry(False)
-                    return {
-                        "msg": f"Connection error:\n{traceback.format_exc()}",
-                        "result": "connection-error",
-                    }
+                    msg = f"Connection error:\n{traceback.format_exc()}"
+                    if self.legacy_exception_handling:
+                        return {
+                            "msg": msg,
+                            "result": "connection-error",
+                        }
+                    else:
+                        raise UnrecoverableNetworkError(msg)
```

Comment:
Hmm, what does this look like in terms of output?  I'm not sure that we don't want to just reraise `e`.

Reply:
It’s definitely not ideal to raise an exception whose message is the stringified traceback of another exception, unless you’re getting paid per line of traceback.

```pytb
>>> c.get_server_settings()
Traceback (most recent call last):
  File "/home/anders/zulip/python-zulip-api/.direnv/python-3.10.0/lib/python3.10/site-packages/urllib3/connectionpool.py", line 445, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "/home/anders/zulip/python-zulip-api/.direnv/python-3.10.0/lib/python3.10/site-packages/urllib3/connectionpool.py", line 440, in _make_request
    httplib_response = conn.getresponse()
  File "/nix/store/33jhvwkqm1327ypyhzbvl1l5d6ch62jn-python3-3.10.0/lib/python3.10/http/client.py", line 1368, in getresponse
    response.begin()
  File "/nix/store/33jhvwkqm1327ypyhzbvl1l5d6ch62jn-python3-3.10.0/lib/python3.10/http/client.py", line 317, in begin
    version, status, reason = self._read_status()
  File "/nix/store/33jhvwkqm1327ypyhzbvl1l5d6ch62jn-python3-3.10.0/lib/python3.10/http/client.py", line 278, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/nix/store/33jhvwkqm1327ypyhzbvl1l5d6ch62jn-python3-3.10.0/lib/python3.10/socket.py", line 705, in readinto
    return self._sock.recv_into(b)
  File "/nix/store/33jhvwkqm1327ypyhzbvl1l5d6ch62jn-python3-3.10.0/lib/python3.10/ssl.py", line 1273, in recv_into
    return self.read(nbytes, buffer)
  File "/nix/store/33jhvwkqm1327ypyhzbvl1l5d6ch62jn-python3-3.10.0/lib/python3.10/ssl.py", line 1129, in read
    return self._sslobj.read(len, buffer)
TimeoutError: The read operation timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/anders/zulip/python-zulip-api/.direnv/python-3.10.0/lib/python3.10/site-packages/requests/adapters.py", line 439, in send
    resp = conn.urlopen(
  File "/home/anders/zulip/python-zulip-api/.direnv/python-3.10.0/lib/python3.10/site-packages/urllib3/connectionpool.py", line 755, in urlopen
    retries = retries.increment(
  File "/home/anders/zulip/python-zulip-api/.direnv/python-3.10.0/lib/python3.10/site-packages/urllib3/util/retry.py", line 532, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/home/anders/zulip/python-zulip-api/.direnv/python-3.10.0/lib/python3.10/site-packages/urllib3/packages/six.py", line 770, in reraise
    raise value
  File "/home/anders/zulip/python-zulip-api/.direnv/python-3.10.0/lib/python3.10/site-packages/urllib3/connectionpool.py", line 699, in urlopen
    httplib_response = self._make_request(
  File "/home/anders/zulip/python-zulip-api/.direnv/python-3.10.0/lib/python3.10/site-packages/urllib3/connectionpool.py", line 447, in _make_request
    self._raise_timeout(err=e, url=url, timeout_value=read_timeout)
  File "/home/anders/zulip/python-zulip-api/.direnv/python-3.10.0/lib/python3.10/site-packages/urllib3/connectionpool.py", line 336, in _raise_timeout
    raise ReadTimeoutError(
urllib3.exceptions.ReadTimeoutError: HTTPSConnectionPool(host='anders-test.zulipchat.com', port=443): Read timed out. (read timeout=15.0)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/anders/zulip/python-zulip-api/zulip/zulip/__init__.py", line 629, in do_api_query
    res = self.session.request(
  File "/home/anders/zulip/python-zulip-api/.direnv/python-3.10.0/lib/python3.10/site-packages/requests/sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/anders/zulip/python-zulip-api/.direnv/python-3.10.0/lib/python3.10/site-packages/requests/sessions.py", line 655, in send
    r = adapter.send(request, **kwargs)
  File "/home/anders/zulip/python-zulip-api/.direnv/python-3.10.0/lib/python3.10/site-packages/requests/adapters.py", line 529, in send
    raise ReadTimeout(e, request=request)
requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='anders-test.zulipchat.com', port=443): Read timed out. (read timeout=15.0)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/anders/zulip/python-zulip-api/zulip/zulip/__init__.py", line 1117, in get_server_settings
    return self.call_endpoint(
  File "/home/anders/zulip/python-zulip-api/zulip/zulip/__init__.py", line 727, in call_endpoint
    return self.do_api_query(
  File "/home/anders/zulip/python-zulip-api/zulip/zulip/__init__.py", line 665, in do_api_query
    raise UnrecoverableNetworkError(msg)
zulip.UnrecoverableNetworkError: Connection error:
Traceback (most recent call last):
  File "/home/anders/zulip/python-zulip-api/.direnv/python-3.10.0/lib/python3.10/site-packages/urllib3/connectionpool.py", line 445, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "/home/anders/zulip/python-zulip-api/.direnv/python-3.10.0/lib/python3.10/site-packages/urllib3/connectionpool.py", line 440, in _make_request
    httplib_response = conn.getresponse()
  File "/nix/store/33jhvwkqm1327ypyhzbvl1l5d6ch62jn-python3-3.10.0/lib/python3.10/http/client.py", line 1368, in getresponse
    response.begin()
  File "/nix/store/33jhvwkqm1327ypyhzbvl1l5d6ch62jn-python3-3.10.0/lib/python3.10/http/client.py", line 317, in begin
    version, status, reason = self._read_status()
  File "/nix/store/33jhvwkqm1327ypyhzbvl1l5d6ch62jn-python3-3.10.0/lib/python3.10/http/client.py", line 278, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/nix/store/33jhvwkqm1327ypyhzbvl1l5d6ch62jn-python3-3.10.0/lib/python3.10/socket.py", line 705, in readinto
    return self._sock.recv_into(b)
  File "/nix/store/33jhvwkqm1327ypyhzbvl1l5d6ch62jn-python3-3.10.0/lib/python3.10/ssl.py", line 1273, in recv_into
    return self.read(nbytes, buffer)
  File "/nix/store/33jhvwkqm1327ypyhzbvl1l5d6ch62jn-python3-3.10.0/lib/python3.10/ssl.py", line 1129, in read
    return self._sslobj.read(len, buffer)
TimeoutError: The read operation timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/anders/zulip/python-zulip-api/.direnv/python-3.10.0/lib/python3.10/site-packages/requests/adapters.py", line 439, in send
    resp = conn.urlopen(
  File "/home/anders/zulip/python-zulip-api/.direnv/python-3.10.0/lib/python3.10/site-packages/urllib3/connectionpool.py", line 755, in urlopen
    retries = retries.increment(
  File "/home/anders/zulip/python-zulip-api/.direnv/python-3.10.0/lib/python3.10/site-packages/urllib3/util/retry.py", line 532, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/home/anders/zulip/python-zulip-api/.direnv/python-3.10.0/lib/python3.10/site-packages/urllib3/packages/six.py", line 770, in reraise
    raise value
  File "/home/anders/zulip/python-zulip-api/.direnv/python-3.10.0/lib/python3.10/site-packages/urllib3/connectionpool.py", line 699, in urlopen
    httplib_response = self._make_request(
  File "/home/anders/zulip/python-zulip-api/.direnv/python-3.10.0/lib/python3.10/site-packages/urllib3/connectionpool.py", line 447, in _make_request
    self._raise_timeout(err=e, url=url, timeout_value=read_timeout)
  File "/home/anders/zulip/python-zulip-api/.direnv/python-3.10.0/lib/python3.10/site-packages/urllib3/connectionpool.py", line 336, in _raise_timeout
    raise ReadTimeoutError(
urllib3.exceptions.ReadTimeoutError: HTTPSConnectionPool(host='anders-test.zulipchat.com', port=443): Read timed out. (read timeout=15.0)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/anders/zulip/python-zulip-api/zulip/zulip/__init__.py", line 629, in do_api_query
    res = self.session.request(
  File "/home/anders/zulip/python-zulip-api/.direnv/python-3.10.0/lib/python3.10/site-packages/requests/sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/anders/zulip/python-zulip-api/.direnv/python-3.10.0/lib/python3.10/site-packages/requests/sessions.py", line 655, in send
    r = adapter.send(request, **kwargs)
  File "/home/anders/zulip/python-zulip-api/.direnv/python-3.10.0/lib/python3.10/site-packages/requests/adapters.py", line 529, in send
    raise ReadTimeout(e, request=request)
requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='anders-test.zulipchat.com', port=443): Read timed out. (read timeout=15.0)
```

I agree that reraising the error directly (`raise`) is a good initial solution. If we ever need a more specific error so it can be caught somewhere else, it should be `raise SpecificError("succinct one-line message") from e` (see [exception chaining](https://docs.python.org/3/tutorial/errors.html#exception-chaining)).

Reply:
Agreed! :)